docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,528 @@
+# MeshExternalService API
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6330
+
+## Context and Problem Statement
+
+The current implementation of `ExternalService` has significant limitations and warrants a complete redesign.
+
+* Lack of support for * subdomains
+* Lack of support for thick clients (think Kafka or C* where discovery is done as part of the client).
+* Need for pluggability (for example lambda support or equivalent)
+* Working well with gateways (what if your ExternalService relies on SNI/host header to route through a gateway)
+* Applying policies with and without egress
+
+
+## Considered Options
+
+* Creating MeshExternalService + HostnameGenerator
+* Creating MeshExternalService
+* Creating MeshExternalService and using HostnameGenerator for InternalVIP
+
+## Decision Outcome
+
+Chosen option: "Creating MeshExternalService".
+
+### Positive Consequences
+
+* Clearer structure.
+* Enhanced functionalities.
+* Better policy matching capabilities.
+* No need for a HostnameGenerator
+
+### Negative Consequences
+
+* Whole new resource with a lot of new code.
+* Temporarily increased complexity of a product until the migration is done.
+* Problem with supporting Universal without transparent proxy, because one MeshExternalService points to many real domains
+
+## Pros and Cons of the Options
+
+### Creating MeshExternalService
+
+We're introducing a new object resource `MeshExternalService`. This object will be a namespace resource and will only be creatable within the `kuma-system` namespace. It can be created within a specific zone, but in this case, it will only be accessible within the zone where it was created.
```

Comment:
Would it be synced to the Global-CP which is created in the zone-CP (https://kuma.io/docs/2.7.x/policies/applying-policies/#applying-policies-on-zone-and-global-control-planes).

And if could, should we use a ValidatingWebhook to forbid the Global-CP deleting operation? Actually, I did a test with `MeshTrafficPermission` and found I can delete the Global-CP `MeshTrafficPermission` copy while the zone-cp one still exists.

Reply:
It should be synced for a GUI purpose. We should have a separate task to figure out how to behave in this case. If we should forbid this

Reply:
@Icarus9913 I think what you pointed is: https://github.com/kumahq/kuma/issues/10260